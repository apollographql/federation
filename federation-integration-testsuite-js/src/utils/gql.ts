import { parse, Kind } from "graphql";
import stripIndent from 'strip-indent';
// This function signature is mostly lifted from the upstream `graphql-tag`.
// The motivation for this `gql` implementation is the desire to preserve
// the `locations` attributes from parsing, which `graphql-tag` purges.
// We don't typically use the function-invocation (e.g., `gql("string")`),
// but the same mode is preserved from `graphql-tag`.  The similarities
// might be appreciated.  (Or not, but having symmetry in a same-named
// function is somewhat appealing, indeed.)
export function gql(
  literals: string | readonly string[],
  ...args: any[]
) {

  if (typeof literals === 'string') {
    literals = [literals];
  }

  let result = literals[0];

  args.forEach((arg, i) => {
    if (arg && arg.kind === Kind.DOCUMENT) {
      result += arg.loc.source.body;
    } else {
      result += arg;
    }
    result += literals[i + 1];
  });

  return parse(stripIndent(result));
}
