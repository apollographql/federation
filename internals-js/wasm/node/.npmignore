# We don't want to check in build artifacts (thus the .gitignore), but we _do_ need NPM to publish those files
# .npmignore overrides .gitignore