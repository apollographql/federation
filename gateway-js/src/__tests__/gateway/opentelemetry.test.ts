import gql from 'graphql-tag';
import {ApolloGateway, LocalGraphQLDataSource} from '../../';
import {fixtures, spanSerializer} from 'apollo-federation-integration-testsuite';
import {InMemorySpanExporter, SimpleSpanProcessor} from '@opentelemetry/tracing'
import {NodeTracerProvider} from '@opentelemetry/node';
import { buildSubgraphSchema } from '@apollo/subgraph';

expect.addSnapshotSerializer(spanSerializer);

const inMemorySpans = new InMemorySpanExporter();
const tracerProvider = new NodeTracerProvider();
tracerProvider.addSpanProcessor(new SimpleSpanProcessor(inMemorySpans));
tracerProvider.register();

beforeEach(() => {
  inMemorySpans.reset();
});

describe('opentelemetry', () => {
  async function execute(executor: any, source: string, variables: any, operationName: string) {
    await executor({
      source,
      document: gql(source),
      request: {
        variables,
      },
      operationName,
      queryHash: 'hashed',
      context: null,
      cache: {} as any,
    });
  }

  describe('with local data', () =>
  {
    async function gateway() {
      const localDataSources = Object.fromEntries(
        fixtures.map((f) => [
          f.name,
          new LocalGraphQLDataSource(buildSubgraphSchema(f)),
        ]),
      );
      const gateway = new ApolloGateway({
        localServiceList: fixtures,
        buildService(service) {
          return localDataSources[service.name];
        },
      });

      const {executor} = await gateway.load();
      return executor;
    }

    it('receives spans on success', async () => {
      const executor = await gateway();

      const source = `#graphql
      query GetProduct($upc: String!) {
        product(upc: $upc) {
          name
        }
      }
    `;

      await execute(executor, source, {upc: '1'}, 'GetProduct');
      expect(inMemorySpans.getFinishedSpans()).toMatchSnapshot();
    });

    it('receives spans on validation failure', async () => {
      const executor = await gateway();
      const source = `#graphql
      query InvalidVariables($first: Int!) {
        topReviews(first: $first) {
          body
        }
      }
    `;

      await execute(executor, source, { upc: '1' }, 'InvalidVariables');
      expect(inMemorySpans.getFinishedSpans()).toMatchSnapshot();
    });

    it('receives spans on plan failure', async () => {
      const executor = await gateway();
      const source = `#graphql
      subscription GetProduct($upc: String!) {
        product(upc: $upc) {
          name
        }
      }
    `;

      try {
        await execute(executor, source, {upc: '1'}, 'GetProduct');
      }
      catch(err) {}
      expect(inMemorySpans.getFinishedSpans()).toMatchSnapshot();
    });
  });


  it('receives spans on fetch failure', async () => {
    const gateway = new ApolloGateway({
      localServiceList: fixtures,
      fetcher: () => {
        throw Error('Nooo');
      },
    });

    const { executor } = await gateway.load();

    const source = `#graphql
    query GetProduct($upc: String!) {
      product(upc: $upc) {
        name
      }
    }
    `;

    await execute(executor, source, {upc: '1'}, 'GetProduct');
    expect(inMemorySpans.getFinishedSpans()).toMatchSnapshot();
  });
});
