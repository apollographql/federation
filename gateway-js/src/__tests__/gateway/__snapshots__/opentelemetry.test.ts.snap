// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`opentelemetry receives spans on fetch failure 1`] = `
[
  {
    "name": "gateway.request",
    "attributes": {
      "operationName": "GetProduct"
    },
    "children": [
      {
        "name": "gateway.validate",
        "attributes": {},
        "children": [],
        "status": {
          "code": 0
        }
      },
      {
        "name": "gateway.plan",
        "attributes": {},
        "children": [],
        "status": {
          "code": 0
        }
      },
      {
        "name": "gateway.execute",
        "attributes": {},
        "children": [
          {
            "name": "gateway.fetch",
            "attributes": {
              "service": "product"
            },
            "children": [],
            "status": {
              "code": 2
            }
          },
          {
            "name": "gateway.postprocessing",
            "attributes": {},
            "children": [],
            "status": {
              "code": 0
            }
          }
        ],
        "status": {
          "code": 2
        }
      }
    ],
    "status": {
      "code": 2
    }
  }
]
`;

exports[`opentelemetry with local data receives spans on plan failure 1`] = `
[
  {
    "name": "gateway.request",
    "attributes": {
      "operationName": "GetProduct"
    },
    "children": [
      {
        "name": "gateway.validate",
        "attributes": {},
        "children": [],
        "status": {
          "code": 0
        }
      },
      {
        "name": "gateway.plan",
        "attributes": {},
        "children": [],
        "status": {
          "code": 2
        }
      }
    ],
    "status": {
      "code": 2
    }
  }
]
`;

exports[`opentelemetry with local data receives spans on success 1`] = `
[
  {
    "name": "gateway.request",
    "attributes": {
      "operationName": "GetProduct"
    },
    "children": [
      {
        "name": "gateway.validate",
        "attributes": {},
        "children": [],
        "status": {
          "code": 0
        }
      },
      {
        "name": "gateway.plan",
        "attributes": {},
        "children": [],
        "status": {
          "code": 0
        }
      },
      {
        "name": "gateway.execute",
        "attributes": {},
        "children": [
          {
            "name": "gateway.fetch",
            "attributes": {
              "service": "product"
            },
            "children": [],
            "status": {
              "code": 0
            }
          },
          {
            "name": "gateway.fetch",
            "attributes": {
              "service": "books"
            },
            "children": [],
            "status": {
              "code": 0
            }
          },
          {
            "name": "gateway.fetch",
            "attributes": {
              "service": "product"
            },
            "children": [],
            "status": {
              "code": 0
            }
          },
          {
            "name": "gateway.postprocessing",
            "attributes": {},
            "children": [],
            "status": {
              "code": 0
            }
          }
        ],
        "status": {
          "code": 0
        }
      }
    ],
    "status": {
      "code": 0
    }
  }
]
`;

exports[`opentelemetry with local data receives spans on validation failure 1`] = `
[
  {
    "name": "gateway.request",
    "attributes": {
      "operationName": "InvalidVariables"
    },
    "children": [
      {
        "name": "gateway.validate",
        "attributes": {},
        "children": [],
        "status": {
          "code": 2
        }
      }
    ],
    "status": {
      "code": 2
    }
  }
]
`;
